File: android/src/main/java/app/capgo/capacitor/intercom/CapgoIntercomPlugin.java
package app.capgo.capacitor.intercom;

import android.app.Application;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.intercom.android.sdk.Company;
import io.intercom.android.sdk.Intercom;
import io.intercom.android.sdk.IntercomSpace;
import io.intercom.android.sdk.UnreadConversationCountListener;
import io.intercom.android.sdk.UserAttributes;
import io.intercom.android.sdk.identity.Registration;
import io.intercom.android.sdk.push.IntercomPushClient;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "CapgoIntercom")
public class CapgoIntercomPlugin extends Plugin {

    private static final String TAG = "CapgoIntercom";
    private final IntercomPushClient intercomPushClient = new IntercomPushClient();
    private UnreadConversationCountListener unreadListener;

    @Override
    public void load() {
        try {
            String apiKey = getConfig().getString("androidApiKey", "");
            String appId = getConfig().getString("androidAppId", "");

            if (apiKey != null && !apiKey.isEmpty() && appId != null && !appId.isEmpty()) {
                Application app = (Application) getContext().getApplicationContext();
                Intercom.initialize(app, apiKey, appId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Intercom: " + e.getMessage(), e);
        }

        unreadListener = (unreadCount) -> {
            JSObject data = new JSObject();
            data.put("count", unreadCount);
            notifyListeners("unreadCountDidChange", data);
        };
        Intercom.client().addUnreadConversationCountListener(unreadListener);
    }

    @Override
    public void handleOnStart() {
        try {
            String apiKey = getConfig().getString("androidApiKey", "");
            String appId = getConfig().getString("androidAppId", "");

            if (apiKey != null && !apiKey.isEmpty() && appId != null && !appId.isEmpty()) {
                Application app = (Application) getContext().getApplicationContext();
                Intercom.initialize(app, apiKey, appId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to re-initialize Intercom on start: " + e.getMessage(), e);
        }
    }

    @Override
    protected void handleOnDestroy() {
        if (unreadListener != null) {
            Intercom.client().removeUnreadConversationCountListener(unreadListener);
            unreadListener = null;
        }
    }

    @PluginMethod
    public void loadWithKeys(PluginCall call) {
        String apiKey = call.getString("apiKeyAndroid", "");
        String appId = call.getString("appId", "");

        if (apiKey == null || apiKey.isEmpty() || appId == null || appId.isEmpty()) {
            call.reject("appId and apiKeyAndroid are required");
            return;
        }

        try {
            Application app = (Application) getContext().getApplicationContext();
            Intercom.initialize(app, apiKey, appId);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to initialize Intercom: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void registerIdentifiedUser(PluginCall call) {
        String userId = call.getString("userId");
        String email = call.getString("email");

        Registration registration = new Registration();
        if (userId != null && !userId.isEmpty()) {
            registration = registration.withUserId(userId);
        }
        if (email != null && !email.isEmpty()) {
            registration = registration.withEmail(email);
        }

        Intercom.client().registerIdentifiedUser(registration);
        call.resolve();
    }

    @PluginMethod
    public void registerUnidentifiedUser(PluginCall call) {
        Intercom.client().registerUnidentifiedUser();
        call.resolve();
    }

    @PluginMethod
    public void updateUser(PluginCall call) {
        UserAttributes.Builder builder = new UserAttributes.Builder();

        String userId = call.getString("userId");
        if (userId != null && !userId.isEmpty()) {
            builder.withUserId(userId);
        }

        String email = call.getString("email");
        if (email != null && !email.isEmpty()) {
            builder.withEmail(email);
        }

        String name = call.getString("name");
        if (name != null && !name.isEmpty()) {
            builder.withName(name);
        }

        String phone = call.getString("phone");
        if (phone != null && !phone.isEmpty()) {
            builder.withPhone(phone);
        }

        String languageOverride = call.getString("languageOverride");
        if (languageOverride != null && !languageOverride.isEmpty()) {
            builder.withLanguageOverride(languageOverride);
        }

        JSObject customAttributes = call.getObject("customAttributes");
        if (customAttributes != null) {
            Map<String, Object> map = mapFromJSON(customAttributes);
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                builder.withCustomAttribute(entry.getKey(), entry.getValue());
            }
        }

        try {
            JSONArray companiesArray = call.getArray("companies");
            if (companiesArray != null) {
                for (int i = 0; i < companiesArray.length(); i++) {
                    JSONObject companyData = companiesArray.getJSONObject(i);
                    Company.Builder companyBuilder = new Company.Builder();

                    if (companyData.has("companyId")) {
                        companyBuilder.withCompanyId(companyData.getString("companyId"));
                    }
                    if (companyData.has("name")) {
                        companyBuilder.withName(companyData.getString("name"));
                    }
                    if (companyData.has("plan")) {
                        companyBuilder.withPlan(companyData.getString("plan"));
                    }
                    if (companyData.has("monthlySpend")) {
                        companyBuilder.withMonthlySpend(companyData.getInt("monthlySpend"));
                    }
                    if (companyData.has("createdAt")) {
                        companyBuilder.withCreatedAt(companyData.getLong("createdAt"));
                    }
                    if (companyData.has("customAttributes")) {
                        JSONObject customAttrs = companyData.getJSONObject("customAttributes");
                        Map<String, Object> attrsMap = mapFromJSON(JSObject.fromJSONObject(customAttrs));
                        for (Map.Entry<String, Object> entry : attrsMap.entrySet()) {
                            companyBuilder.withCustomAttribute(entry.getKey(), entry.getValue());
                        }
                    }

                    builder.withCompany(companyBuilder.build());
                }
            }
        } catch (JSONException e) {
            Log.w(TAG, "Failed to parse companies: " + e.getMessage());
        }

        Intercom.client().updateUser(builder.build());
        call.resolve();
    }

    @PluginMethod
    public void logout(PluginCall call) {
        Intercom.client().logout();
        call.resolve();
    }

    @PluginMethod
    public void logEvent(PluginCall call) {
        String name = call.getString("name");
        if (name == null || name.isEmpty()) {
            call.reject("Event name is required");
            return;
        }

        JSObject data = call.getObject("data");
        if (data != null) {
            Map<String, Object> metaData = mapFromJSON(data);
            Intercom.client().logEvent(name, metaData);
        } else {
            Intercom.client().logEvent(name);
        }
        call.resolve();
    }

    @PluginMethod
    public void displayMessenger(PluginCall call) {
        Intercom.client().present();
        call.resolve();
    }

    @PluginMethod
    public void displayMessageComposer(PluginCall call) {
        String message = call.getString("message");
        if (message == null) {
            call.reject("message is required");
            return;
        }
        Intercom.client().displayMessageComposer(message);
        call.resolve();
    }

    @PluginMethod
    public void displayHelpCenter(PluginCall call) {
        Intercom.client().present(IntercomSpace.HelpCenter);
        call.resolve();
    }

    @PluginMethod
    public void hideMessenger(PluginCall call) {
        Intercom.client().hideIntercom();
        call.resolve();
    }

    @PluginMethod
    public void displayLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.Visibility.VISIBLE);
        call.resolve();
    }

    @PluginMethod
    public void hideLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.Visibility.GONE);
        call.resolve();
    }

    @PluginMethod
    public void displayInAppMessages(PluginCall call) {
        Intercom.client().setInAppMessageVisibility(Intercom.Visibility.VISIBLE);
        call.resolve();
    }

    @PluginMethod
    public void hideInAppMessages(PluginCall call) {
        Intercom.client().setInAppMessageVisibility(Intercom.Visibility.GONE);
        call.resolve();
    }

    @PluginMethod
    public void displayCarousel(PluginCall call) {
        String carouselId = call.getString("carouselId");
        if (carouselId == null || carouselId.isEmpty()) {
            call.reject("carouselId is required");
            return;
        }
        Intercom.client().displayCarousel(carouselId);
        call.resolve();
    }

    @PluginMethod
    public void displayArticle(PluginCall call) {
        String articleId = call.getString("articleId");
        if (articleId == null || articleId.isEmpty()) {
            call.reject("articleId is required");
            return;
        }
        Intercom.client().displayArticle(articleId);
        call.resolve();
    }

    @PluginMethod
    public void displaySurvey(PluginCall call) {
        String surveyId = call.getString("surveyId");
        if (surveyId == null || surveyId.isEmpty()) {
            call.reject("surveyId is required");
            return;
        }
        Intercom.client().displaySurvey(surveyId);
        call.resolve();
    }

    @PluginMethod
    public void setUserHash(PluginCall call) {
        String hmac = call.getString("hmac");
        if (hmac == null || hmac.isEmpty()) {
            call.reject("hmac is required");
            return;
        }
        try {
            Intercom.client().setUserHash(hmac);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to set user hash: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void setUserJwt(PluginCall call) {
        String jwt = call.getString("jwt");
        if (jwt == null || jwt.isEmpty()) {
            call.reject("jwt is required");
            return;
        }
        try {
            Intercom.client().setUserHash(jwt);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to set user JWT: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void setBottomPadding(PluginCall call) {
        Double value = call.getDouble("value");
        if (value == null) {
            call.reject("value is required");
            return;
        }
        Intercom.client().setBottomPadding(value.intValue());
        call.resolve();
    }

    @PluginMethod
    public void sendPushTokenToIntercom(PluginCall call) {
        String token = call.getString("value");
        if (token == null || token.isEmpty()) {
            call.reject("value is required");
            return;
        }
        intercomPushClient.sendTokenToIntercom(getActivity().getApplication(), token);
        call.resolve();
    }

    @PluginMethod
    public void receivePush(PluginCall call) {
        JSObject notificationData = call.getData();
        if (notificationData != null) {
            Map<String, String> map = new HashMap<>();
            Iterator<String> keys = notificationData.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                try {
                    map.put(key, notificationData.getString(key));
                } catch (Exception e) {
                    // skip non-string values
                }
            }
            if (intercomPushClient.isIntercomPush(map)) {
                intercomPushClient.handlePush(getActivity().getApplication(), map);
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void getUnreadConversationCount(PluginCall call) {
        int count = Intercom.client().getUnreadConversationCount();
        JSObject result = new JSObject();
        result.put("count", count);
        call.resolve(result);
    }

    private static Map<String, Object> mapFromJSON(JSObject jsonObject) {
        Map<String, Object> map = new HashMap<>();
        Iterator<String> keys = jsonObject.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            try {
                Object value = jsonObject.get(key);
                if (value instanceof JSONObject) {
                    map.put(key, mapFromJSON(JSObject.fromJSONObject((JSONObject) value)));
                } else if (value instanceof JSONArray) {
                    map.put(key, listFromJSON((JSONArray) value));
                } else {
                    map.put(key, value);
                }
            } catch (JSONException e) {
                // skip
            }
        }
        return map;
    }

    private static List<Object> listFromJSON(JSONArray jsonArray) {
        List<Object> list = new ArrayList<>();
        for (int i = 0; i < jsonArray.length(); i++) {
            try {
                Object value = jsonArray.get(i);
                if (value instanceof JSONObject) {
                    list.add(mapFromJSON(JSObject.fromJSONObject((JSONObject) value)));
                } else if (value instanceof JSONArray) {
                    list.add(listFromJSON((JSONArray) value));
                } else {
                    list.add(value);
                }
            } catch (JSONException e) {
                // skip
            }
        }
        return list;
    }
}

File: ios/Sources/CapgoIntercomPlugin/CapgoIntercomPlugin.swift
import Foundation
import Capacitor
import Intercom

@objc(CapgoIntercomPlugin)
public class CapgoIntercomPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapgoIntercomPlugin"
    public let jsName = "CapgoIntercom"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "loadWithKeys", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "registerIdentifiedUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "registerUnidentifiedUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayMessenger", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayMessageComposer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayHelpCenter", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideMessenger", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayLauncher", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideLauncher", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayInAppMessages", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideInAppMessages", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayCarousel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayArticle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displaySurvey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserHash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserJwt", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setBottomPadding", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendPushTokenToIntercom", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "receivePush", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getUnreadConversationCount", returnType: CAPPluginReturnPromise)
    ]

    private var showObserver: NSObjectProtocol?
    private var hideObserver: NSObjectProtocol?
    private var pushObserver: NSObjectProtocol?
    private var unreadObserver: NSObjectProtocol?

    override public func load() {
        let apiKey = getConfig().getString("iosApiKey") ?? ""
        let appId = getConfig().getString("iosAppId") ?? ""

        if !apiKey.isEmpty && !appId.isEmpty {
            Intercom.setApiKey(apiKey, forAppId: appId)
            #if DEBUG
            Intercom.enableLogging()
            #endif
        }

        pushObserver = NotificationCenter.default.addObserver(
            forName: .capacitorDidRegisterForRemoteNotifications,
            object: nil,
            queue: .main
        ) { notification in
            if let deviceToken = notification.object as? Data {
                Intercom.setDeviceToken(deviceToken, completion: nil)
            }
        }

        showObserver = NotificationCenter.default.addObserver(
            forName: NSNotification.Name("IntercomWindowDidShowNotification"),
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.notifyListeners("windowDidShow", data: [:])
        }

        hideObserver = NotificationCenter.default.addObserver(
            forName: NSNotification.Name("IntercomWindowDidHideNotification"),
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.notifyListeners("windowDidHide", data: [:])
        }

        unreadObserver = NotificationCenter.default.addObserver(
            forName: NSNotification.Name.IntercomUnreadConversationCountDidChange,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            let count = Intercom.unreadConversationCount()
            self?.notifyListeners("unreadCountDidChange", data: ["count": count])
        }
    }

    deinit {
        if let obs = showObserver {
            NotificationCenter.default.removeObserver(obs)
        }
        if let obs = hideObserver {
            NotificationCenter.default.removeObserver(obs)
        }
        if let obs = pushObserver {
            NotificationCenter.default.removeObserver(obs)
        }
        if let obs = unreadObserver {
            NotificationCenter.default.removeObserver(obs)
        }
    }

    @objc func loadWithKeys(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? ""
        let apiKey = call.getString("apiKeyIOS") ?? ""

        guard !appId.isEmpty && !apiKey.isEmpty else {
            call.reject("appId and apiKeyIOS are required")
            return
        }

        Intercom.setApiKey(apiKey, forAppId: appId)
        #if DEBUG
        Intercom.enableLogging()
        #endif
        call.resolve()
    }

    @objc func registerIdentifiedUser(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let email = call.getString("email")

        let attributes = ICMUserAttributes()
        if let userId = userId {
            attributes.userId = userId
        }
        if let email = email {
            attributes.email = email
        }

        Intercom.loginUser(with: attributes) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject("Failed to register user: \(error.localizedDescription)")
            }
        }
    }

    @objc func registerUnidentifiedUser(_ call: CAPPluginCall) {
        Intercom.loginUnidentifiedUser { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject("Failed to register unidentified user: \(error.localizedDescription)")
            }
        }
    }

    @objc func updateUser(_ call: CAPPluginCall) {
        let attributes = ICMUserAttributes()

        if let userId = call.getString("userId") {
            attributes.userId = userId
        }
        if let email = call.getString("email") {
            attributes.email = email
        }
        if let name = call.getString("name") {
            attributes.name = name
        }
        if let phone = call.getString("phone") {
            attributes.phone = phone
        }
        if let languageOverride = call.getString("languageOverride") {
            attributes.languageOverride = languageOverride
        }
        if let customAttributes = call.getObject("customAttributes") {
            attributes.customAttributes = customAttributes as [String: Any]
        }

        if let companiesArray = call.getArray("companies") as? [[String: Any]] {
            var companies: [ICMCompany] = []
            for companyData in companiesArray {
                let company = ICMCompany()
                if let companyId = companyData["companyId"] as? String {
                    company.companyId = companyId
                }
                if let name = companyData["name"] as? String {
                    company.name = name
                }
                if let plan = companyData["plan"] as? String {
                    company.plan = plan
                }
                if let monthlySpend = companyData["monthlySpend"] as? NSNumber {
                    company.monthlySpend = monthlySpend
                }
                if let createdAt = companyData["createdAt"] as? Double {
                    company.createdAt = Date(timeIntervalSince1970: createdAt)
                }
                if let customAttrs = companyData["customAttributes"] as? [String: Any] {
                    company.customAttributes = customAttrs
                }
                companies.append(company)
            }
            attributes.companies = companies
        }

        Intercom.updateUser(with: attributes)
        call.resolve()
    }

    @objc func logout(_ call: CAPPluginCall) {
        Intercom.logout()
        call.resolve()
    }

    @objc func logEvent(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("Event name is required")
            return
        }

        if let data = call.getObject("data") {
            Intercom.logEvent(withName: name, metaData: data as [AnyHashable: Any])
        } else {
            Intercom.logEvent(withName: name)
        }
        call.resolve()
    }

    @objc func displayMessenger(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.present()
            call.resolve()
        }
    }

    @objc func displayMessageComposer(_ call: CAPPluginCall) {
        guard let message = call.getString("message") else {
            call.reject("message is required")
            return
        }

        DispatchQueue.main.async {
            Intercom.presentMessageComposer(message)
            call.resolve()
        }
    }

    @objc func displayHelpCenter(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.present(.helpCenter)
            call.resolve()
        }
    }

    @objc func hideMessenger(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.hide()
            call.resolve()
        }
    }

    @objc func displayLauncher(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setLauncherVisible(true)
            call.resolve()
        }
    }

    @objc func hideLauncher(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setLauncherVisible(false)
            call.resolve()
        }
    }

    @objc func displayInAppMessages(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setInAppMessagesVisible(true)
            call.resolve()
        }
    }

    @objc func hideInAppMessages(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setInAppMessagesVisible(false)
            call.resolve()
        }
    }

    @objc func displayCarousel(_ call: CAPPluginCall) {
        guard let carouselId = call.getString("carouselId") else {
            call.reject("carouselId is required")
            return
        }

        DispatchQueue.main.async {
            Intercom.presentContent(.carousel(id: carouselId))
            call.resolve()
        }
    }

    @objc func displayArticle(_ call: CAPPluginCall) {
        guard let articleId = call.getString("articleId") else {
            call.reject("articleId is required")
            return
        }

        DispatchQueue.main.async {
            Intercom.presentContent(.article(id: articleId))
            call.resolve()
        }
    }

    @objc func displaySurvey(_ call: CAPPluginCall) {
        guard let surveyId = call.getString("surveyId") else {
            call.reject("surveyId is required")
            return
        }

        DispatchQueue.main.async {
            Intercom.presentContent(.survey(id: surveyId))
            call.resolve()
        }
    }

    @objc func setUserHash(_ call: CAPPluginCall) {
        guard let hmac = call.getString("hmac") else {
            call.reject("hmac is required")
            return
        }

        Intercom.setUserHash(hmac)
        call.resolve()
    }

    @objc func setUserJwt(_ call: CAPPluginCall) {
        guard let jwt = call.getString("jwt") else {
            call.reject("jwt is required")
            return
        }

        Intercom.setUserHash(jwt)
        call.resolve()
    }

    @objc func setBottomPadding(_ call: CAPPluginCall) {
        guard let value = call.getDouble("value") else {
            call.reject("value is required")
            return
        }

        DispatchQueue.main.async {
            Intercom.setBottomPadding(CGFloat(value))
            call.resolve()
        }
    }

    @objc func sendPushTokenToIntercom(_ call: CAPPluginCall) {
        guard let token = call.getString("value") else {
            call.reject("value is required")
            return
        }

        if let tokenData = token.data(using: .utf8) {
            DispatchQueue.main.async {
                Intercom.setDeviceToken(tokenData, completion: nil)
            }
        }
        call.resolve()
    }

    @objc func receivePush(_ call: CAPPluginCall) {
        let userInfo = call.options as? [AnyHashable: Any] ?? [:]
        DispatchQueue.main.async {
            if Intercom.isIntercomPushNotification(userInfo) {
                Intercom.handlePushNotification(userInfo)
            }
        }
        call.resolve()
    }

    @objc func getUnreadConversationCount(_ call: CAPPluginCall) {
        let count = Intercom.unreadConversationCount()
        call.resolve(["count": count])
    }
}
